File: server/src/main/java/ch/zxseitz/tbsg/server/websocket/GameSocketHandler.java
package ch.zxseitz.tbsg.server.websocket;

import ch.zxseitz.tbsg.TbsgException;
import ch.zxseitz.tbsg.games.GameState;
import ch.zxseitz.tbsg.games.IGame;
import ch.zxseitz.tbsg.games.exceptions.ClientException;
import ch.zxseitz.tbsg.server.games.GameManager;
import ch.zxseitz.tbsg.server.games.GameProxy;

import java.util.*;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.socket.CloseStatus;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;
import org.springframework.web.socket.handler.TextWebSocketHandler;

public class GameSocketHandler extends TextWebSocketHandler {
    private final Logger logger;
    private final GameProxy proxy;
    private final Map<String, Client> clients;

    public GameSocketHandler(GameProxy proxy) {
        this.logger = LoggerFactory.getLogger(GameSocketHandler.class.getName() + ":" + proxy.getName());
        this.proxy = proxy;
        this.clients = new ConcurrentHashMap<>();
    }

    /**
     * Performs a critical section on several locks
     *
     * @param callable critical section
     * @param locks    list of lock
     * @return critical section return value
     * @throws Exception if an exception occurs during the critical section
     */
    @SafeVarargs
    private <T, L> T safe(Callable<T> callable, IProtectable<L>... locks) throws Exception {
        var sorted = Arrays.stream(locks).sorted().collect(Collectors.toList());  //prevent deadlocks
        sorted.forEach(IProtectable::lock);
        try {
            return callable.call();
        } finally {
            sorted.forEach(IProtectable::unlock);
        }
    }

    /**
     * Handles client connection
     *
     * @param session client websocket session
     */
    @Override
    public void afterConnectionEstablished(WebSocketSession session) throws Exception {
        logger.info("Connected client: {}", session.getId());
        // todo: read auth infos
        var client = new Client(session);
        clients.put(session.getId(), client);
        sendToClient(client, MessageManager.createIdMessage(client));
    }

    /**
     * Handles client disconnection
     *
     * @param session client websocket session
     * @param status  close status
     */
    @Override
    public void afterConnectionClosed(WebSocketSession session, CloseStatus status) {
        logger.info("Disconnected client: {}", session.getId());
        var client = clients.remove(session.getId());
        // todo delete challenges and abort open matches
    }

    /**
     * Handles client events
     *
     * @param session client websocket session
     * @param message client event
     */
    @Override
    public void handleTextMessage(WebSocketSession session, TextMessage message) {
        var client = clients.get(session.getId());

        try {
            var codeNode = MessageManager.parseClientMessage(message.getPayload());
            var messageCode = codeNode.getKey();
            var argNode = codeNode.getValue();
            // lobby client events
            switch (messageCode) {
                case MessageManager.CLIENT_CHALLENGE: {
                    var opponentId = MessageManager.readClientArgument(argNode, "opponent", String.class);
                    var opponent = clients.get(opponentId);
                    if (opponent == null) {
                        throw new TbsgException(String.format("Opponent [%s] is not connected", opponentId));
                    }
                    safe(() -> {
                        client.getChallenges().add(opponent);
                        opponent.send(MessageManager.createChallengeMessage(client));
                        return null;
                    }, client);
                    break;
                }
                case MessageManager.CLIENT_CHALLENGE_ABORT: {
                    var opponentId = MessageManager.readClientArgument(argNode, "opponent", String.class);
                    var opponent = clients.get(opponentId);
                    if (opponent == null) {
                        throw new TbsgException(String.format("Opponent [%s] is not connected", opponentId));
                    }
                    safe(() -> {
                        if (client.getChallenges().remove(opponent)) {
                            opponent.send(MessageManager.createChallengeAbortMessage(client));
                        } else {
                            throw new TbsgException(String.format("Opponent [%s] is not challenged by you", opponentId));
                        }
                        return null;
                    }, client);
                    break;
                }
                case MessageManager.CLIENT_CHALLENGE_ACCEPT: {
                    var opponentId = MessageManager.readClientArgument(argNode, "opponent", String.class);
                    var opponent = clients.get(opponentId);
                    if (opponent == null) {
                        throw new TbsgException(String.format("Opponent [%s] is not connected", opponentId));
                    }
                    safe(() -> {
                        // todo queuing accepts
                        // fixme color values
                        if (client.getMatch() == null && opponent.getMatch() == null) {
                            if (client.getChallenges().remove(opponent)) {
                                opponent.send(MessageManager.createChallengeAcceptMessage(client));
                                var game = proxy.createGame();
                                var clients = new TreeMap<Integer, Client>();
                                clients.put(1, client);
                                clients.put(2, opponent);
                                var match = GameManager.createMatch(game, clients);
                                client.setMatch(match);
                                opponent.setMatch(match);

                                sendToClient(client, MessageManager
                                        .createGameInitMessage(2, game.getBoard()));
                                sendToClient(opponent, MessageManager
                                        .createGameInitNextMessage(1, game.getBoard(), game.getPreview()));
                            } else {
                                throw new TbsgException(String.format("Opponent [%s] is not challenged by you", opponentId));
                            }
                        } else {
                            throw new TbsgException(String.format("You or opponent [%s] is currently in game", opponentId));
                        }
                        return null;
                    }, client, opponent);
                    break;
                }
                case MessageManager.CLIENT_CHALLENGE_DECLINE: {
                    var opponentId = MessageManager
                            .readClientArgument(argNode, "opponent", String.class);
                    var opponent = clients.get(opponentId);
                    if (opponent == null) {
                        throw new TbsgException(String.format("Opponent [%s] is not connected", opponentId));
                    }
                    safe(() -> {
                        if (opponent.getChallenges().remove(client)) {
                            opponent.send(MessageManager.createChallengeDeclineMessage(client));
                        } else {
                            throw new TbsgException(String.format("Opponent [%s] is not challenged by you", opponentId));
                        }
                        return null;
                    }, opponent);
                    break;
                }
                case MessageManager.CLIENT_UPDATE: {
                    var action = MessageManager
                            .readClientGameArguments(argNode, proxy.getActionClass());
                    safe(() -> {
                        var match = client.getMatch();
                        var gameProtector = match.getGame();
                        var opponent = match.getOpponent(client);
                        safe(() -> {
                            var game = (IGame<Object>) gameProtector.get();
                            if (game.getState() == GameState.FINISHED) {
                                sendToClient(client, MessageManager
                                        .createErrorMessage("Game is already finished"));
                            }
                            var color = match.getColor(client);
                            if (game.getNext() != color) {
                                sendToClient(client, MessageManager
                                        .createErrorMessage("Not your turn"));
                            }

                            game.update(action);

                            if (game.getState().equals(GameState.RUNNING)) {
                                // game continues
                                var updateNextMessage = MessageManager
                                        .createGameUpdateNextMessage(action, game.getBoard(), game.getPreview());
                                var updateMessage = MessageManager
                                        .createGameUpdateMessage(action, game.getBoard());
                                if (game.getNext() == color) {
                                    sendToClient(client, updateNextMessage);
                                    sendToClient(opponent, updateMessage);
                                } else {
                                    sendToClient(client, updateMessage);
                                    sendToClient(opponent, updateNextMessage);
                                }
                            } else {
                                // game finished
                                var next = game.getNext();
                                if (next == 0) {
                                    var tieMessage = MessageManager
                                            .createGameEndTieMessage(action, game.getBoard());
                                    sendToClient(client, tieMessage);
                                    sendToClient(opponent, tieMessage);
                                } else {
                                    var victoryMessage = MessageManager
                                            .createGameEndVictoryMessage(action, game.getBoard());
                                    var defeatMessage = MessageManager
                                            .createGameEndDefeatMessage(action, game.getBoard());
                                    if (next == color) {
                                        sendToClient(client, victoryMessage);
                                        sendToClient(opponent, defeatMessage);
                                    } else {
                                        sendToClient(client, defeatMessage);
                                        sendToClient(opponent, victoryMessage);
                                    }
                                }
                            }
                            return null;
                        }, gameProtector);
                        return null;
                    }, client);
                }
                default: {
                    sendToClient(client, MessageManager.createErrorMessage("unknown message code "
                            + messageCode));
                }
            }
        } catch (Exception e) {
            logger.warn(e.getMessage(), e);
            try {
                //todo cache and try resending, recognize new client
                sendToClient(client, MessageManager.createErrorMessage(e.getMessage()));
            } catch (Exception ce) {
                logger.warn(ce.getMessage(), ce);
            }
        }
    }

    private void sendToClient(Client client, String message) {
        try {
            client.send(message);
        } catch (ClientException ce) {
            logger.warn(ce.getMessage());
        }
    }
}

File: server/src/main/java/ch/zxseitz/tbsg/server/websocket/GameSocketHandler.kt
package ch.zxseitz.tbsg.server.websocket

import ch.zxseitz.tbsg.TbsgException
import ch.zxseitz.tbsg.games.GameState
import ch.zxseitz.tbsg.games.IGame
import ch.zxseitz.tbsg.games.exceptions.ClientException
import ch.zxseitz.tbsg.games.exceptions.GameException
import ch.zxseitz.tbsg.server.games.GameManager
import ch.zxseitz.tbsg.server.games.GameProxy

import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.web.socket.CloseStatus
import org.springframework.web.socket.TextMessage
import org.springframework.web.socket.WebSocketSession
import org.springframework.web.socket.handler.TextWebSocketHandler
import java.util.*

class GameSocketHandler(private val proxy: GameProxy): TextWebSocketHandler() {
    private val logger: Logger = LoggerFactory.getLogger(GameSocketHandler::class.java.name + ":" + proxy.name)
    private val clients: MutableMap<String, Client> = ConcurrentHashMap()

    /**
     * Performs a critical section on several locks
     *
     * @param callable critical section
     * @param locks    list of lock
     * @return critical section return value
     * @throws Exception if an exception occurs during the critical section
     */
    private fun <T, L> safe(callable: () -> T, vararg locks: IProtectable<L>): T {
        val sorted = Arrays.stream(locks).sorted().collect(Collectors.toList())  //prevent deadlocks
        sorted.forEach(IProtectable<L>::lock)
        try {
            return callable()
        } finally {
            sorted.forEach(IProtectable<L>::unlock)
        }
    }

    /**
     * Handles client connection
     *
     * @param session client websocket session
     */
    override fun afterConnectionEstablished(session: WebSocketSession) {
        logger.info("Connected client: ${session.id}")
        // todo: read auth infos
        val client = Client(session)
        clients[session.id] = client
        sendToClient(client, createIdMessage(client))
    }

    /**
     * Handles client disconnection
     *
     * @param session client websocket session
     * @param status  close status
     */
    override fun afterConnectionClosed(session: WebSocketSession, status: CloseStatus) {
        logger.info("Disconnected client: ${session.id}")
        var client = clients.remove(session.id)
        // todo delete challenges and abort open matches
    }

    /**
     * Handles client events
     *
     * @param session client websocket session
     * @param message client event
     */
    public override fun handleTextMessage(session: WebSocketSession, message: TextMessage) {
        val client = clients[session.id]!!

        try {
            val codeNode = parseClientMessage(message.payload)
            val messageCode = codeNode.first
            val argNode = codeNode.second
            // lobby client events
            when (messageCode) {
                CLIENT_CHALLENGE -> {
                    val opponentId = readClientArgument(argNode, "opponent", String::class.java)
                    val opponent = clients[opponentId] ?: throw TbsgException("Opponent [$opponentId] is not connected")
                    safe({
                        client.challenges.add(opponent)
                        opponent.send(createChallengeMessage(client))
                    }, client)
                }
                CLIENT_CHALLENGE_ABORT -> {
                    val opponentId = readClientArgument(argNode, "opponent", String::class.java)
                    val opponent = clients[opponentId] ?: throw TbsgException("Opponent [$opponentId] is not connected")
                    safe({
                        if (client.challenges.remove(opponent)) {
                            opponent.send(createChallengeAbortMessage(client))
                        } else {
                            throw TbsgException("Opponent [$opponentId] is not challenged by you")
                        }
                    }, client)
                }
                CLIENT_CHALLENGE_ACCEPT -> {
                    val opponentId = readClientArgument(argNode, "opponent", String::class.java)
                    val opponent = clients[opponentId] ?: throw TbsgException("Opponent [$opponentId] is not connected")
                    safe({
                        // todo queuing accepts
                        // fixme color values
                        if (client.match == null && opponent.match == null) {
                            if (opponent.challenges.remove(client)) {
                                opponent.send(createChallengeAcceptMessage(client))
                                val game = proxy.createGame()
                                val clients = TreeMap<Int, Client>()
                                clients[1] = client
                                clients[2] = opponent
                                val match = GameManager.createMatch(game, clients)
                                client.match = match
                                opponent.match = match
                                game.init()

                                sendToClient(client, createGameInitNextMessage(1, game.board, game.preview))
                                sendToClient(opponent, createGameInitMessage(2, game.board))
                            } else {
                                throw TbsgException("Opponent [$opponentId] is not challenged by you")
                            }
                        } else {
                            throw TbsgException("You or opponent [$opponentId] is currently in game")
                        }
                    }, client, opponent)
                }
                CLIENT_CHALLENGE_DECLINE -> {
                    val opponentId = readClientArgument(argNode, "opponent", String::class.java)
                    val opponent = clients[opponentId] ?: throw TbsgException("Opponent [$opponentId] is not connected")
                    safe({
                        if (opponent.challenges.remove(client)) {
                            opponent.send(createChallengeDeclineMessage(client))
                        } else {
                            throw TbsgException("Opponent [$opponentId] is not challenged by you")
                        }
                    }, opponent)
                }
                CLIENT_UPDATE -> {
                    val action = readClientGameArguments(argNode, proxy.actionClass)
                    safe({
                        val match = client.match!!
                        val gameProtector = match.game
                        val opponent = match.getOpponent(client)
                        safe({
                            val game = gameProtector.t
                            if (game.state == GameState.FINISHED) {
                                throw GameException("Game is already finished")
                            }
                            val color = match.getColor(client)
                            if (game.next != color) {
                                throw GameException("Not your turn")
                            }

                            game.update(action)

                            if (game.state == GameState.RUNNING) {
                                // game continues
                                val updateNextMessage = createGameUpdateNextMessage(action, game.board, game.preview)
                                val updateMessage = createGameUpdateMessage(action, game.board)
                                if (game.next == color) {
                                    sendToClient(client, updateNextMessage)
                                    sendToClient(opponent, updateMessage)
                                } else {
                                    sendToClient(client, updateMessage)
                                    sendToClient(opponent, updateNextMessage)
                                }
                            } else {
                                // game finished
                                val next = game.next
                                if (next == 0) {
                                    val tieMessage = createGameEndTieMessage(action, game.board)
                                    sendToClient(client, tieMessage)
                                    sendToClient(opponent, tieMessage)
                                } else {
                                    val victoryMessage = createGameEndVictoryMessage(action, game.board)
                                    val defeatMessage = createGameEndDefeatMessage(action, game.board)
                                    if (next == color) {
                                        sendToClient(client, victoryMessage)
                                        sendToClient(opponent, defeatMessage)
                                    } else {
                                        sendToClient(client, defeatMessage)
                                        sendToClient(opponent, victoryMessage)
                                    }
                                }
                            }
                        }, gameProtector)
                    }, client)
                }
                else -> {
                    sendToClient(client, createErrorMessage("unknown message code: $messageCode"))
                }
            }
        } catch (e: Exception) {
//            logger.warn(e.message, e)
            logger.warn(e.message)
            sendToClient(client, createErrorMessage(e.message!!))
        }
    }

    private fun sendToClient(client: Client, message: String) {
        try {
            //todo cache and try resending, recognize new client
            client.send(message)
        } catch (ce: ClientException) {
            logger.warn(ce.message)
        }
    }
}
